### ISSUE TITLE
[Linux] Error 307 from vrcompositor segfault

### ISSUE THREAD

**Issue by CFSworks** [alvr-org/ALVR#1427]
## Description
When launching SteamVR with ALVR enabled, the `vrcompositor` fails to launch. A segfault is logged in the kernel log:
`[ 5978.237812] vrcompositor.re[32989]: segfault at 10 ip 00007f4d659a06f0 sp 00007ffdb21f3368 error 4 in libc.so.6[7f4d6593e000+14b000] likely on CPU 13 (core 5, socket 0)`

Studying the coredump reveals that the segfault is from `vr::CHmdWindowSDL::AcquireNextImage(..., 2)` (in `vrcompositor`) calling `pthread_mutex_lock(nullptr);`

This problem seemed to appear on its own without me making any changes to the system: when I first set up ALVR ~3 days ago, everything worked fine. Yesterday, I started having crashes from #1315. Building latest `master` from source, `v19.0.0` w/ #1324 cherry-picked, or using precompiled `v19.0.0` w/ `wsi_layer_vkGetPhysicalDeviceDisplayPropertiesKHR` binary-patched by hand resolves #1315 but lets `vrcompositor` initialize far enough to encounter this crash instead.

No ALVR functions seem to be in the coredump's call stack, so I can't be 100% sure that this isn't an issue with SteamVR itself. It does not happen when I launch SteamVR without ALVR enabled, but I don't have any hardware HMDs connected, so it may simply be that ALVR allows SteamVR to begin attempting to AcquireNextImage.

Also, since this problem started happening seemingly without provocation, I suspect it might come and go on its own. I might be unable to keep troubleshooting if the problem stops occurring reliably.

## General Troubleshooting
- [X] I carefully followed the instructions in the [README](https://github.com/alvr-org/ALVR/blob/master/README.md) and successfully completed the setup wizard
- [X] I read the ALVR Wikis [here](https://github.com/polygraphene/ALVR/wiki) and [here](https://github.com/alvr-org/ALVR/wiki)

### Hardware

**CPU**: `11th Gen Intel(R) Core(TM) i9-11900H @ 2.50GHz`

**GPU**: `NVIDIA GeForce RTX 3080 Laptop GPU/PCIe/SSE2`

**Audio**: N/A

### Installation
**ALVR Version**: Today's `master` (from source) and v19.0.0 (prebuilt `alvr_server_linux.tar.gz`)

**SteamVR Version**: 1671750570

**Install Type**:
- [ ] Packaged (`exe`, `deb`, `rpm`, etc)
- [X] Portable (`zip`)
- [X] Source

**OS Name and Version**: Gentoo

**Comment by CFSworks** [comment]
By the way, my suspicions just proved correct: I left to go do something else for a while and when I came back, everything started working fine again. So, it seems like this is one of those "it depends on its mood at the moment" faults.

**Comment by Vixea** [comment]
Please try the latest release and get make to us about it

**Comment by CFSworks** [comment]
Hi! I have tried 19.1.0 (the AppImage this time as the release omits the `alvr_server_linux.tar.gz` package) and it has the same behavior.

I spent some time troubleshooting and found that the error only occurs when `/dev/nvidia-modeset` is missing. I have a hardened system that apparently does not allow `nvidia-modprobe` the privilege to create new devices under `/dev`; this can probably be reproduced on other systems by deleting that device and stopping the nvidia driver from being able to call `nvidia-modprobe`.

Obviously the `/dev/nvidia-modeset` device is an important file to the usermode driver so the VR compositor cannot work without it, however it seems there may be an important error check missing in either ALVR or SteamVR that tries to use the Vulkan device even when it cannot initialize correctly. (This issue should probably be retitled to reflect that, if it is indeed an ALVR issue.)

**Comment by PorkrollPosadist** [comment]
I believe I am experiencing the same bug, however I have an AMD card instead of an Nvidia one.  SteamVR error 307, vrcompositor segfault messages in dmesg.  I tried both v19.1.0 and master (dad4a52f).  I tried loading the core dump in GDB to see how deep the similarities go, but apparently nothing has debug symbols and the backtrace is incomprehensible to me.

The machine is an Intel i7-4770k with 32GB memory and an AMD RX 6700. (Also) running Gentoo with Linux 6.2.7-gentoo-x86_64.  Gnome / Wayland.

**Comment by CFSworks** [comment]
> @CFSworks Great find! Do you think you can make a PR for a fix?

Thank you! I am still not sure if this is an ALVR problem or a SteamVR problem, but if it's an ALVR problem I'll send in a PR if I am able. :)

> I believe I am experiencing the same bug, however I have an AMD card instead of an Nvidia one. SteamVR error 307, vrcompositor segfault messages in dmesg. 

It might or might not be. SteamVR error 307 does appear to be the catch-all "vrcompositor has segfaulted" error code, but if you're seeing a segfaulting access to 0x10 from within libc, it's a fair cop it's related!

> but apparently nothing has debug symbols and the backtrace is incomprehensible to me.

Since you're on Gentoo, I might recommend enabling the `splitdebug` feature in Portage. It eats up 83GB worth of space on my system but I've found it invaluable when I need to do debugging.

Anyway, it looks like the Vulkan call that Nvidia's driver wants `/dev/nvidia-modeset` for is `vkGetPhysicalDeviceSurfacePresentModesKHR`. I might need to do an apitrace or something on ALVR+SteamVR to see what's calling it and if that's indeed the failing call.

@PorkrollPosadist would you be willing to do an apitrace and look for any vk init call that is coming back with an error status?

**Comment by creeloper27** [comment]
> SteamVR error 307 does appear to be the catch-all "vrcompositor has segfaulted" error code

Then my case is weird :/, I am running into the same error 307 on SteamVR tho neither ail -f /var/log/kern.log nor dmesg report any segfaults, the SteamVR Web Console reports:
```
Loaded server driver alvr_server (IServerTrackedDeviceProvider_004) from /home/cattai-lorenzo/Desktop/ALVR/build/alvr_streamer_linux/lib64/alvr/bin/linux64/driver_alvr_server.so
VendorSpecificEvent 111 outside of reserved range
```
I have tried the latest release and latest nightly, even built it from source using master but they all have this behaviour. I am running Pop!_OS 22.04 LTS x86_64 with a NVIDIA GeForce GTX 1050 Mobile running Driver Version: 525.89.02 and CUDA Version: 12.0.
I also tried the latest release, latest nighly and latest build from source using master on my Nobara 37 installation (on the same system) but the behaviour is the same.
I tried on Windows 10 (on the same system) and there everything seems to work fine.
